redis_feature_store.js

```diff
@@ -0,0 +1,159 @@
+  }
+
+  store.all = function(cb) {
+    client.hgetall(features_key, function(err, obj) {
+      if (err) {
+        config.logger.error("[LaunchDarkly] Error fetching flag from redis", err)
+        cb(null);
+      } else {
+        var results = {}, 
+            flags = obj;
+
+        for (var key in flags) {
+          if (Object.hasOwnProperty.call(flags,key)) {
+            var flag = JSON.parse(flags[key]);
+            if (!flag.deleted) {
+              results[key] = flag;          
+            }
+          }
+        }
+        cb(results);
+      }
+    });
+  }
+
+  store.init = function(flags, cb) {
+    var multi = client.multi();
+    
+    multi.del(features_key);
+    cache.clear();
+
+    for (var key in flags) {
+      if (Object.hasOwnProperty.call(flags,key)) {
+        multi.hset(features_key, key, JSON.stringify(flags[key]));
+      }
+      if (cache_ttl) {
+        cache.put(key, flags[key], cache_ttl);
+      }
+    }
+
+    multi.exec(function(err, replies) {
+      if (err) {
+        config.logger.error("[LaunchDarkly] Error initializing redis feature store", err);
+      } 
+      cb();
+    });
+  }
+
+  store.delete = function(key, version, cb) {
+    var multi;
+    client.watch(features_key);
+    multi = client.multi();
+
+    // We need to run the get() code in a multi txn
+    do_get(multi, key, function(flag) {
+      if (flag) {
+        if (flag.version >= version) {
+          cb();
+          return;          
+        } else {
+          flag.deleted = true;
+          flag.version = version;
+          multi.hset(features_key, key, JSON.stringify(flag));
+          multi.exec(function(err, replies) {
+            if (err) {
+              config.logger.error("[LaunchDarkly] Error deleting feature flag", err);
+            } else {
+              cache.put(key, flag, cache_ttl);
+            }
+            cb();
+          })
+        }
+      } 
+    });
+  }
+
+  store.upsert = function(key, flag, cb) {
+    var multi;
+    client.watch(features_key);
+    multi = client.multi();
+
+    do_get(multi, key, function(original) {
+      if (original) {
+        if (original.version >= version) {
+          cb();
+          return;          
+        } else {
+          multi.hset(features_key, key, JSON.stringify(flag));
+          multi.exec(function(err, replies) {
+            if (err) {
+              config.logger.error("[LaunchDarkly] Error upserting feature flag", err);
+            } else {
+              cache.put(key, flag, cache_ttl);
+            }
+            cb();
+          })
+        }
+      } 
+    })
+  }
+
+  store.initialized = function(cb) {
+    var init = cache.get('$initialized$');
+
+    if (init) {
+      return true;
+    }    
+
+    client.exists('$initialized$', function(err, obj) {
+      if (!err && obj) {
+        cache.set('$initialized$', true);
+      } 
+      cb(!err && obj);
+    });
+  }
+
+  store.close = function() {
+    client.quit();
```

Comment:
I'm having a problem with the test driver when the Redis feature store is configured-- I can't get the program to exit cleanly, even if I call `client.close()`. Still investigating.